Replumbing: add support for multiple names per algorithm

levitte opened the pull request:
Algorithms may have multiple names, as seen in the legacy names
database.  We need to support that as well.

This implementations modifies ossl_namemap to support multiple names
for the same identifier.

levitte commented:
Note that I haven't tested at all, only made sure it builds on my machine.
To be finished tomorrow.

levitte commented:
@paulidale, this is a good spot to continue our discussion from #8728

beldmit commented:
Do we need to think about a case when we have the same name for different algorithms provided by different providers? It can happen by accident or because of misunderstanding.

slontis commented on crypto/core_namemap.c:
```diff
@@ -30,6 +40,7 @@ struct ossl_namemap_st {
     CRYPTO_RWLOCK *lock;
     LHASH_OF(NAMEMAP_ENTRY) *namenum;  /* Name->number mapping */
     STACK_OF(NAMEMAP_ENTRY) *numname;  /* Number->name mapping */
+    SPARSE_ARRAY_OF(ALIAS_LIST) *numnum; /* Number->number alias table */
```
Not sure about this variable name :)

levitte replied:
Not sure I want to read your mind ;-)

slontis commented:
Wouldnt it be better to have aliases in libcrypto so they can be shared between providers..
It should still be possible to add these aliases into each provider when they are registered.
If a provider wants it own special case aliases then we would also need a list of aliases on the provider side also (Not necessarily embedded in the algorithm table).

levitte commented:
> Do we need to think about a case when we have the same name for different algorithms provided by different providers? It can happen by accident or because of misunderstanding.

That is always a possibility. In current terms, there's nothing stopping anyone from creating an engine that pretends to implement SHA1, but that does something else. I'm counting on the ecosystem of engines/providers to weed itself out of such "bad apples" (in practical terms, it's about the reliability and popularity if each module, on the responsibility of their diverse authors).

levitte commented:
> Wouldnt it be better to have aliases in libcrypto so they can be shared between providers..
> It should still be possible to add these aliases into each provider when they are registered.

I don't see why it's better to centralize aliases in libcrypto. Thing is that the namemap *is* shared within a library context, so any alias given from a provider will end up being "global". 

However, this made it obvious that the code in this PR isn't quite right. If several providers implement the same algo, but with a slightly different (but still overlapping) set of names, they may not all be added (first such provider wins), or may still be seen as different algorithms (because first given name in each doesn't match). I will fix that later today.

paulidale commented on providers/common/ciphers/aes.c:
```diff
@@ -443,35 +444,36 @@ static int aes_ctx_set_params(void *vctx, const OSSL_PARAM params[])
-IMPLEMENT_block_funcs(cbc, 192, 16)
-IMPLEMENT_block_funcs(cbc, 128, 16)
+IMPLEMENT_block_funcs(cbc, 256, 0) = { "AES-256-CBC", "aes-256-cbc", NULL };
+IMPLEMENT_block_funcs(cbc, 192, 0) = { "AES-192-CBC", "aes-192-cbc", NULL };
```
The 16 values have been replaced with 0 here but not below?
It doesn't look like the value is used in the macros but consistency seems worthwhile.

mattcaswell replied:
The values *is* used...

````
        { OSSL_FUNC_CIPHER_IV_LENGTH, (void (*)(void))iv_length_##ivlen }, \
````

levitte replied:
Oops, the 0s are wrong, of course...  Will fix

levitte commented:
Rethinking where aliases come from, it's true that no one stops us from pre-populating the namemap if we wish... I'll explore that in another PR.

paulidale commented:
This still feels like the wrong place to do the aliasing.  It seems like it is adding complexity (a layer of indirection) that isn't required.

I'm wondering if we need this at all?  For the provider internal EVP layer (i.e. FIPS), aliases aren't required.  All resolutions are fixed.  That leaves libcrypto's EVP interface which handles aliasing already.

We will need an ability for a provider to register a new algorithm name with aliases.  No disputing that.

A libcrypto call to convert a name into a canonical form might be worthwhile exporting to providers.


If we do want aliases inside the providers, the core name map can be modified to handle multiple names mapping onto a single integer and this complexity is avoided.  The reverse map from integer to name could be to the canonical form (or first registered or whatever).

BTW: I think we're also risking the loss of case insensitive look up of algorithm names.

levitte commented:
> This still feels like the wrong place to do the aliasing. It seems like it is adding complexity (a layer of indirection) that isn't required.

Not sure I follow. Aliases are names, right? So how can it be wrong to handle that where names are handled? 

> I'm wondering if we need this at all? For the provider internal EVP layer (i.e. FIPS), aliases aren't required. All resolutions are fixed. That leaves libcrypto's EVP interface which handles aliasing already.

No it doesn't (apart from case insensitivity). That's why the namemap came to be, so we'd avoid involving the object database. 

> We will need an ability for a provider to register a new algorithm name with aliases. No disputing that.

> A libcrypto call to convert a name into a canonical form might be worthwhile exporting to providers.

What would the provider do with that? Also, what is "canonical form"? I'm actually trying to get away from the idea that there is *one true name*, and rather see that *all* names for an algorithm are "equal". 

> If we do want aliases inside the providers, the core name map can be modified to handle multiple names mapping onto a single integer and this complexity is avoided. The reverse map from integer to name could be to the canonical form (or first registered or whatever).

That's what I am trying to do on the API level. The internal structure is a what it is, and it needs being reworked to allow adding more names for the same algo (i.e needs a bit more flexibility) 
BTW: I think we're also risking the loss of case insensitive look up of algorithm names.

Deano1167 commented:
God said only one key would be given over all others.  This is where you
all have an offline do you have it no do you no.  Buhahaha

On Tue, May 21, 2019, 8:29 PM Richard Levitte <notifications@github.com>
wrote:

> This still feels like the wrong place to do the aliasing. It seems like it
> is adding complexity (a layer of indirection) that isn't required.
>
> Not sure I follow. Aliases are names, right? So how can it be wrong to
> handle that where names are handled?
>
> I'm wondering if we need this at all? For the provider internal EVP layer
> (i.e. FIPS), aliases aren't required. All resolutions are fixed. That
> leaves libcrypto's EVP interface which handles aliasing already.
>
> No it doesn't (apart from case insensitivity). That's why the namemap came
> to be, so we'd avoid involving the object database.
>
> We will need an ability for a provider to register a new algorithm name
> with aliases. No disputing that.
>
> A libcrypto call to convert a name into a canonical form might be
> worthwhile exporting to providers.
>
> What would the provider do with that? Also, what is "canonical form"? I'm
> actually trying to get away from the idea that there is *one true name*,
> and rather see that *all* names for an algorithm are "equal".
>
> If we do want aliases inside the providers, the core name map can be
> modified to handle multiple names mapping onto a single integer and this
> complexity is avoided. The reverse map from integer to name could be to the
> canonical form (or first registered or whatever).
>
> That's what I am trying to do on the API level. The internal structure is
> a what it is, and it needs being reworked to allow adding more names for
> the same algo (i.e needs a bit more flexibility)
> BTW: I think we're also risking the loss of case insensitive look up of
> algorithm names.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openssl/openssl/pull/8967?email_source=notifications&email_token=AJ4SQIOTQJPL6PYALETJON3PWSOZJA5CNFSM4HOLS772YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV5UE5I#issuecomment-494617205>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJ4SQIKLO75ITXTXYOWM32LPWSOZJANCNFSM4HOLS77Q>
> .
>

paulidale commented:
> That's what I am trying to do on the API level. The internal structure is a what it is, and it needs being reworked to allow adding more names for the same algo (i.e needs a bit more flexibility)

What about adding:
```
int ossl_namemap_alias(OSSL_NAMEMAP *namemap, int number, const char *alias)
```
or maybe allow an array of aliases instead of one?


The usage being:
```
int n = ossl_namemap_add(namemap, "foo");
ossl_name_map_alias(namemap, n, "bar");
```

paulidale commented:
> Not sure I follow. Aliases are names, right? So how can it be wrong to handle that where names are handled?

I'm starting to wonder where OIDs will be dealt with?  They are essentially names of a different format.  I'm assuming libcrypto will map an OID to exactly one name.  A provider will need to register that name or face badness.  This all smells like canonical.

I think this also implies `objects.txt` will remain in libcrypto and thus the algorithm aliases are known to libcrypto.  Providers don't need to know about them if so.

levitte commented:
objects.txt isn't the only source of aliases. It's main intention is to map OIDs to names, and I would say that either way, a provider that wants to have its implemented reached through an OID will have to use a matching name (among all the names they can use). 

This is obviously not just a technical discussion, but also one of principle: who shall have the power and the responsibility to decide what names an algorithm implementation shall have? My vote goes to providers.

paulidale commented:
Perhaps a conference call is in order about this?

levitte commented:
> Perhaps a conference call is in order about this?

Sure.

richsalz commented:
Let us think of the users and application writers.  If aliases are part of the provider, then it is quite possible that different systems could have conflicting aliases.  For example, someone may use GOST as an alias, whereas someone else will use the more official name GOST-2018 or some such.  This means that applications could break at run time when moved to a different system.  That would be very bad.

*For this release,* it seems that the simplest way to prevent this from happening is to have aliases only defined and supported in the core. We would be hoping that providers only use standard or official names, and that those names are "enforced" by standards organizations, nations, or other third-party. This does not completely solve the problem, but it is probably good enough for a first release of the module architecture, and it has worked well enough so far in OpenSSL history; look at the comparatively small number of aliases.

You can always add this later. You cannot remove it later if added now.  Wait to see what developers of providers need and want, first.

levitte commented:
So if I understand correctly, we do want multiple names (i.e aliases) *somehow*. So, right, if not implemented as part of the namemap API (which is majorly what this PR is about), then how? Remember that we're banning asn1 code and thereby the object db from the fips module.

It seems to me that this effort is needed somehow, even if we don't give the providers the opportunity to specify multiple names. 
(BTW, if we are to give them the possibility some day, how do you propose to do that? It can obviously not be via the algorithm specification structure, 'cause that will break the API, unless we make that change before releasing 3.0)

richsalz commented:
Maybe you do want multiple names. My suggestion is you wait until after you get some experience and interaction with the community and see what they want/need, not just the couple of people from the OMC in this discussion.

FWIW, my preference would be for an API providers could call to add aliases for official names, and to support that in the config file.  But, again, *not for this release.*

levitte commented:
@richsalz, you said you wanted aliases defined in libcrypto. That can actually done by pre-populating appropriate namemaps... Or are you suggesting another mechanism?
I'm frankly kind of lost by what you're saying, 'cause you want some kind of alias support (in libcrypto at the very least), but I don't see what mechanics you want to be able to provide that.

richsalz commented:
I guess I was using shorthand that could be misleading; sorry about that.

Aliases should be defined in the core, as part of the release.  Just the way they are now.  Do not let providers add their own aliases; they can only use a single new name (and you have to cross your fingers and hope it's an official name defined somewhere), or they can use an existing name defined by OpenSSL.  [That argues there should be a way for a provider to query the core if a crypto mechanism name is known or not.]

Later on, after there's more experience and a better understanding of what is actually needed, then allow run-time aliasing, provider-set aliases, etc.

Is that more clear?

levitte commented:
> Aliases should be defined in the core, as part of the release. *Just the way they are now.*

"as they are now" means using the object db, which means dragging in asn1 code, which we try to avoid.

levitte commented:
So uhmmm, are you saying that `crypto/evp/evp_fetch.c` should use the object database when it isn't built for the FIPS module, and namemap when it's built for the FIPS module?  Is that what you're getting at?

I'm not looking forward to the added `FIPS_MODE` guards, I gotta say that.

richsalz commented:
Can this not be handled in the same way that crypto_malloc, etc., are done?
Or pull the object database stuff out of the ASN1 code.  Want me to try that?

levitte commented:
The object database uses ASN1_OBJECT, and that's part of its API...

beldmit commented:
> > Aliases should be defined in the core, as part of the release. _Just the way they are now._
> 
> "the way they are now" means using the object db, which means dragging in asn1 code, which we try to avoid.

Now I can write smth like 
```
EVP_add_digest_alias(SN_id_GostR3411_2012_256, "streebog256");
EVP_add_digest_alias(SN_id_GostR3411_2012_512, "streebog512");
```
in engine, and it will work.

richsalz commented:
@beldmit  and if we had upcalls to the core, then this could work in providers as well.
@levitte and separating the names from ASN1-object is what your PR does, I get it. I would go further and move all names out of the ASN1 stuff if possible.

levitte commented:
> I would go further and move all names out of the ASN1 stuff if possible.

I have said quite a few times now that it's perfectly possible to prepopulate the namemap.  What you just said there is exactly what that would accomplish.

richsalz commented:
Again, to be clear: I don't think this should be done now.  Wait for more input from users and developers.

levitte commented:
I'm giving up this discussion, it's going in circles

richsalz commented:
Sorry you feel that way.  The points I was making are: don't do it now, but when you do it, consider doing it this way.  Doesn't seem circular to me.

richsalz commented:
Adding a clarification that @levitte and I worked out via email.

This PR has two parts: having a name database with aliases in libcrypto, and letting providers define aliases.

I was assuming that you can't remove the first part as it breaks compatibility.  I do not believe the second part should be done in the 3.0 release.

levitte committed:
Replumbing: add support for multiple names per algorithm

Algorithms may have multiple names, as seen in the legacy names
database.  We need to support that as well.

This implementations modifies ossl_namemap to support multiple names
for the same identifier.

levitte commented:
I have reduced this PR to have the namemap changes only.  I'll submit 2 new PRs for pre-populated as well as for multiple name support in providers.

slontis commented:
Is this still a draft?

slontis commented:
Maybe add some tests for the API's?

levitte commented:
> Maybe add some tests for the API's?

Sure, will do in the next few days

slontis commented on include/internal/namemap.h:
```diff
@@ -16,6 +16,6 @@ OSSL_NAMEMAP *ossl_namemap_stored(OPENSSL_CTX *libctx);
-int ossl_namemap_add(OSSL_NAMEMAP *namemap, const char *name);
-const char *ossl_namemap_name(const OSSL_NAMEMAP *namemap, int number);
+int ossl_namemap_add(OSSL_NAMEMAP *namemap, int number, const char *name);
+const char **ossl_namemap_names(const OSSL_NAMEMAP *namemap, int number);
```
Alternate names?
ossl_namemap_id2names
ossl_namemap_name2id

levitte replied:
That's a though...  although, I would prefer 'num' over 'id'

levitte committed:
Add a namemap test

levitte commented:
Test added!

paulidale commented on doc/internal/man3/ossl_namemap_new.pod:
```diff
@@ -3,7 +3,7 @@
 
 ossl_namemap_new, ossl_namemap_free, ossl_namemap_stored,
-ossl_namemap_add, ossl_namemap_name, ossl_namemap_number
+ossl_namemap_add, ossl_namemap_num2names, ossl_namemap_name2num
```
Just wondering what the use case for `ossl_namemap_num2names` is?

I'm having difficulty thinking of a reasonable one.  If this function is omitted, the code can be significantly simplified.

The best I can come up with is to print the names correspond to an ID but that shouldn't be time critical and iterating over the hash table is probably acceptable.

levitte replied:
I can imagine that down the line, it could help applications print the set of names for an algorithm. I believe we have a lister of ciphers that does this with the help of the object db... Of course, that will require a bit of wrapping as these functions aren't directly available to the application, but still...

richsalz replied:
"I can imagine that down the line, it could help"

I strongly encourage that this release only include what is needed.  You can always add things later, once you have a better understanding of the use-case and requirements. You cannot remove things, and modifying things if it's not quite right is difficult.

levitte replied:
This is an *internal* API, so removing what turns out to be unused is very easy.

paulidale replied:
The point was that a less complex implementation is possible if this is missing -- removing an internal API is easy, remember to rewrite the code into the simpler form isn't.

For printing, the slow path of iterating through all elements of the name to int hash table seems reasonable enough.

levitte commented:
Because popular demand, I've modified this to have a slower iterator over all names corresponding to a number instead of a complex structure holding them.

Re-review, please?

levitte commented:
... and then, #8984 and #8985 need a look

levitte commented:
Oh, and looking at the CI stuff now.

paulidale commented on crypto/core_namemap.c:
```diff
@@ -9,43 +9,48 @@
-static int namemap_cmp(const NAMEMAP_ENTRY *a, const NAMEMAP_ENTRY *b)
+static int namenum_cmp(const NAMENUM_ENTRY *a, const NAMENUM_ENTRY *b)
 {
     return strcmp(a->name, b->name);
```
I still lean towards this using strcasecmp (and openssl_lh_strcasehash above).

Would it make sense to have an extra argument to the new function to allow both case sensitive and insensitive forms?  (or separate new functions for both forms).

levitte replied:
I have nothing against making this case insensitive. We will have a problem if non-ascii strings are ever used, though (well, we already have a problem, but case insensitivity makes for a bigger problem)

levitte replied:
I added another commit that make names case insensitive.

paulidale reviewed (dismissed):
This looks a less complex now.

levitte committed:
OSSL_NAMEMAP: make names case insensitive

levitte commented:
> Travis failure is relevant (doc-nits).

Fixed

levitte commented:
Rebased on fresh master.

levitte commented:
Ping

levitte commented:
Ping

paulidale commented on crypto/core_namemap.c:
```diff
@@ -104,45 +112,43 @@ void ossl_namemap_free(OSSL_NAMEMAP *namemap)
+    if (namenum->number == data->number)
+        data->fn(namenum->name, data->data);
+}
+IMPLEMENT_LHASH_DOALL_ARG_CONST(NAMENUM_ENTRY, DOALL_NAMES_DATA);
```
Blank lines around this?

levitte replied:
Okie

paulidale commented on doc/internal/man3/ossl_namemap_new.pod:
```diff
@@ -38,11 +41,19 @@ ossl_namemap_free().
+ossl_namemap_doall_names() walks through all names associated with
+I<number> in the given I<namemap> and calls the function I<fn> for
+each of them.
+I<fn> gets the I<data> argument as well, which allows any caller to
```
s/gets/is also passed/

levitte replied:
Okie

paulidale reviewed (approved):
A couple of optional nits.

levitte commented:
Merged with last nits fixed.

a9550b74d3efdf1727005dada706cc2e12c273b8 OSSL_NAMEMAP: make names case insensitive
734a462e4028e2f0136d3af0b37611138e781246 Add a namemap test
651d44183e86355b1eb9629a6a61d3da09369bbf Replumbing: add support for multiple names per algorithm
